qiskit_nature/algorithms/initial_points/initial_point.py

```diff
@@ -0,0 +1,66 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The initial point interface."""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+
+import numpy as np
+
+from qiskit_nature.properties.second_quantization import GroupedSecondQuantizedProperty
+from qiskit_nature.circuit.library import UCC
+
+
+class InitialPoint(ABC):
+    """The initial point interface.
+
+    Interface for algorithms that can compute initial points for particular ansatzes.
+    """
+
+    def __init__(self):
+        self._grouped_property: GroupedSecondQuantizedProperty = None
+        self._ansatz: UCC = None
+
+    @property
+    @abstractmethod
+    def initial_point(self) -> np.ndarray:
+        """The initial point."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_initial_point(
+        self, grouped_property: GroupedSecondQuantizedProperty | None, ansatz: UCC | None
+    ) -> np.ndarray:
+        """Computes the initial point."""
+        raise NotImplementedError
```

Comment:
I am not sure if I overlooked this earlier or if this got updated slightly. But from a user perspective seeing both `initial_point` and `get_initial_point` could be a bit confusing, especially since the `initial_point` will fail if `get_initial_point` has not been called earlier.

I would do one of the following:
- rename `get_initial_point -> compute_initial_point` and add a guard inside of `initial_point` to call this if `self._corrections` is None (talking MP2 specific here)
- remove `initial_point`

If you do either of these, the documentation also in MP2 will need updating

Reply:
So, I sort of did both. I didn't really like `InitialPoint.get_initial_point` or `InitialPoint.initial_point`. As it's a bit confusing and verbose. It's a leftover from when the class was briefly called `MP2InformedStart`.

Now if the user tries to call a computed property such as `x` (the actual initial point array) before calling `.compute`, it'll try to call `.compute` with the `grouped_property` and `ansatz` that it already has. If it can't be computed, it'll just raise a warning and return None. I realize that `x` is also a controversial property name 😎, but I felt justified due to the way `x` is used in the [OptimizationResult](https://qiskit.org/documentation/_modules/qiskit/algorithms/optimizers/optimizer.html#OptimizerResult). Still happy to change it, regardless.

Reply:
NB: Refactoring again to use .to_numpy_array(), and more clearly separate computation and conversion.

Reply:
I've now rewritten the API to separate the computation and conversion etc. I've also limited the class properties to those that I expect the user to set and updated all the docs (hopefully I didn't miss any outdated references).